doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+
+Unary minus is arithmetically equivalent to multiplication by minus one; in particular, if _x_ in the expression `-x` is a value with offset type, it is interpreted as a difference quantity, not an absolute quantity, and any offset is ignored.
+
+The square root is introduced by ***square-root*** , and an expression `√x` is arithmetically equivalent to `x^(1/2)`.
+
+The value of an additive, multiplicative or power expression follows the usual calculi of physical quantities (for one of a number of formal treatmes, see for example, P Szekeres, _The mathematical foundations of dimensional analysis and the question of fundamental units_, International Journal of Theoretical Physics 17 no. 12 (1978), doi:10.1007/BF00678423, pp. 957–974).
+
+> _function-application_ ::= ( _qualified-identifier_ | `(` _function-literal_ `)` ) `(` ( _expression_ ( `,` _expression_ )* )? `)`
+
+A _function-application_ expression is well-formed iff the function type of the _function-literal_ or the function value bound to _qualified-identifier_ is compatible with the number and types of the _expression_ clauses constituting the arguments. If the function value corresponds to the form `fn (a₁: T₁, …) → result`, then the value of the function application expression with arguments `expr₁`, … is the value of the expression `let a₁: T₁ = expr₁; … result`.
+
+
+### Record expressions
+<a id="record-expressions"/>
+
+> _record-expr_ ::= _record-term_ ( ***preferential-union*** _record-term_ )*
+>
+> _record-term_ ::= _record-literal_ | _function-application_ | _qualified_identifier_ | _record-literal_ | `(` _record-expr_ `)`
+>
+> _record-field-expr_ ::= ( _record-literal_ | _function-application_ | `(` _record-expr_ `)` ) `.` _qualified-identifier_
+
+Access to a field in a record that is bound to an identifier can be made via the corresponding qualified identifier (see section _Qualified Identifiers_); for records that are not bound to an identifier, the _record-field-expr_ provides the same function, also using the period as an field access operator.
+
+Records of different types can be combined with the ***preferential-union*** operator: given a record _r_ with fields _fᵢ_ and values _vᵢ_ for _i_∈_I_, and another record _s_ with fields _gⱼ_ and values _wⱼ_ for _j_∈_J_, the expression _r_ ⊔ _s_ is a record with a set of fields {_fᵢ_}∪{_gⱼ_}, and the value of a field _h_ in this record is _vᵢ_ if there exists a field _fᵢ_=_h_, else _wⱼ_ for the field _wⱼ_=_h_. Essentially, the record comprises all the fields of the left hand side, together with any fields from the record on the right hand side which don't have the same name as a field in the first record.
```

Comment:
It's there mainly to provide some support for row polymorphism functionality, so we can use a function to 'modify' only some fields of a record. It can potentially save a bunch of tedious binding in a record literal; for a contrived example:
```
def rotate90 = fn (pos: {x: length; y: length;}) → {x = -y; y= x; };
# ...
type point3d = { x: length; y: length; z: length; };
def foo = fn (p: point3d) →
    let p' = rotate90(p) ⊔ p;
    …;
```
as opposed to:
```
def foo = fn (p: point3d) →
    let a = rotate90(p);
    let p' = {x = a.x; y = a.y; z = p.z; };
    …;
```